### PR TITLE
Update DeepSpeech.py

### PR DESCRIPTION
**Pull request by bikramjitroy**
If training happened till 5th epoch then checkpoint file saved as "model-4". Hence start_epoch should be value 5 else 4th epoch checkpoint will be overwritten.

### CODE REVIEW

**Comment by bikramjitroy** [comment]
Changes for training resume should not override the last checkpoint.

**Comment by kdavis-mozilla** [comment]
@bikramjitroy What's your take on this?

**Comment by bikramjitroy** [comment]
I was looking at the code and ran examples below are my observations while running the code.

Environment :  AWS p2 instance with 8 * K80 GPU (p2.8xlarge)
Batch Size: 12 
Dataset Librevox: I have taken 100 hr for training -- Rest all commented in DataSet

**Environment Variable used:**
export ds_importer="librivox"
export ds_train_batch_size=12
export ds_dev_batch_size=12
export ds_test_batch_size=12
export ds_epochs=75
export ds_display_step=1
export ds_checkpoint_step=1
export ds_export_dir="model_export"
export ds_restore_checkpoint=1


Running:
**1)  There is a deadlock in util.spell "correction" API. STacktrace below**

Loaded symbols for /usr/lib/python2.7/lib-dynload/termios.x86_64-linux-gnu.so
__lll_lock_wait_private () at ../nptl/sysdeps/unix/sysv/linux/x86_64/lowlevellock.S:95
95	../nptl/sysdeps/unix/sysv/linux/x86_64/lowlevellock.S: No such file or directory.
(gdb) py-bt
#20 Frame 0x7fe204169410, for file /home/ubuntu/DeepSpeech/util/spell.py, line 21, in log_probability (sentence=['no', 'on', 'on', 'an', 'rs', 'n', 'n', 'land', 'r', 'slot', 'n', 't', 'n', 's', 'r', 'on', 'ours', 'son', 'ol', 'r', 'o', 'n', 'tvs', 'sons', 'or', 'tv', 'rs'])
    return MODEL.score(' '.join(sentence), bos = False, eos = False)
#23 Frame 0x7fe104007c00, for file /home/ubuntu/DeepSpeech/util/spell.py, line 27, in correction (sentence='  no on on rn rs n  n lans r  slst n t n s  r osn oers  son ol r o n tns sons lr   tn rn  w ', layer=[(<float at remote 0xea0f570>, ['no', 'on', 'on', 'an', 'rs', 'n', 'n', 'land', 'r', 'slot', 'n', 't', 'n', 's', 'r', 'on', 'ours', 'son', 'ol', 'r', 'o', 'n', 'tons', 'sons', 'or', 'in']), (<float at remote 0xd312448>, ['no', 'on', 'on', 'an', 'rs', 'n', 'n', 'land', 'r', 'slot', 'n', 't', 'n', 's', 'r', 'on', 'ours', 'son', 'ol', 'r', 'o', 'n', 'tons', 'sons', 'or', 'to']), (<float at remote 0xd1016f8>, ['no', 'on', 'on', 'an', 'rs', 'n', 'n', 'land', 'r', 'slit', 'n', 't', 'n', 's', 'r', 'on', 'ours', 'son', 'ol', 'r', 'o', 'n', 'tons', 'sons', 'or', 'in']), (<float at remote 0x102092c8>, ['no', 'on', 'on', 'an', 'rs', 'n', 'n', 'land', 'r', 'slot', 'n', 't', 'n', 's', 'r', 'on', 'ours', 'son', 'ol', 'r', 'o', 'n', 'tons', 'sons', 'or', 'an']), (<float at remote 0xfb39ab0>, ['no', 'on', 'on', 'an', 'rs', ...(truncated)
    layer = [(-log_probability(node + [cword]), node + [cword]) for cword in candidate_words(word) for priority, node in layer]
#26 Frame 0x7fdfc00086d0, for file DeepSpeech.py, line 613, in calculate_and_print_wer_report (caption='Training loss=415.996428727 avg_cer=0.664574681', results_tuple=(['oh yes he is old man', 'and they buried him in the city of david', 'decision what decision', 'and stayed there for thirty days', 'if it hadnt been for victoria', 'the sky was clear blue', 'saint jamess seven', 'while the two wine bibbers', 'too crowded that is the worst', 'at the same instant the prisoner', 'i will return to morrow', 'the speaker rests his eye', 'northanger abbey', 'missus l h hornsby', 'crayfish soup', 'for beautiful as she now looked', 'my views were in demand', 'you might wonder why i didnt', 'when the rejoicings were over', 'but perhaps the poysers', 'i dont mind im glad youre out of it', 'for love i have heard say', 'that may be answered the boy', 'the instant he made a step farther', 'prawn soup', 'per constructionem', 'psalteries pipes', 'the officiating priest the beadle', 'the story of the grecian king', 'marius beside ...(truncated)
    item = (item[0], correction(item[1]), item[2], item[3])
#29 Frame 0x7fe20808e390, for file DeepSpeech.py, line 994, in print_report (caption='Training', batch_set_result=(<numpy.float64 at remote 0x7fe38c6fd720>, <numpy.float64 at remote 0x7fe38c71c978>, (['oh yes he is old man', 'and they buried him in the city of david', 'decision what decision', 'and stayed there for thirty days', 'if it hadnt been for victoria', 'the sky was clear blue', 'saint jamess seven', 'while the two wine bibbers', 'too crowded that is the worst', 'at the same instant the prisoner', 'i will return to morrow', 'the speaker rests his eye', 'northanger abbey', 'missus l h hornsby', 'crayfish soup', 'for beautiful as she now looked', 'my views were in demand', 'you might wonder why i didnt', 'when the rejoicings were over', 'but perhaps the poysers', 'i dont mind im glad youre out of it', 'for love i have heard say', 'that may be answered the boy', 'the instant he made a step farther', 'prawn soup', 'per constructionem', 'psalteries pipes', 'the officiating priest the beadle', 'the story of the...(truncated)
    mean_wer = calculate_and_print_wer_report(title, results_tuple)
#32 Frame 0x61111a0, for file DeepSpeech.py, line 1085, in train (global_time=<datetime.datetime at remote 0x7fe4eb608e90>, global_train_time=<datetime.timedelta at remote 0x7fe4eb60d3c8>, train_context=(<Graph(_default_original_op=None, _handle_feeders={}, _collections={('__variable_store',): [<_VariableStore(_vars={u'bidirectional_rnn/bw/basic_lstm_cell/biases/Adam_1/': <Variable(_variable=<Tensor(_op=<Operation(_graph=<...>, _control_inputs=[], _outputs=[<...>], _control_flow_context=None, _id_value=12289, _original_op=None, _traceback=[('DeepSpeech.py', 1140, '<module>', {'OrderedDict': <type at remote 0x1b40d40>, 'subprocess': <module at remote 0x7fe4aa62d0c0>, 'collect_results': <function at rem---Type <return> to continue, or q <return> to quit---
ote 0x7fe4e910d6e0>, 'ds_importer': 'librivox', 'run_set': <function at remote 0x7fe4e910dc08>, 'checkpoint_dir': '/home/ubuntu/.local/share/deepspeech', 'get_git_revision_hash': <function at remote 0x7fe4e910d578>, 'importlib': <module at remote 0x7fe4aa62dfa0>, 'tempfile': <module at remote 0x7fe4ab92af68>, 'export...(truncated)
    result = print_report("Training", result)
#35 Frame 0x1a1f180, for file DeepSpeech.py, line 1140, in <module> ()
    last_train_wer, last_dev_wer, hibernation_path = train()
(gdb) exit

**2) BatchSize more than 12 is making the training OOM.**
     Also at batch_size =12 after 3 epoch OOM happens(100% reproducible on p2.8xlarge instance).

**3) Training Time:**
     Every epoch taking 1hr 55 min  (100-hour training data on p2.8xlarge instance).



**Code Observation:** :  
    @:  "print_report" gatting called on every epoch finished. had to comment the code of util.spell "correction" API call to stop spell check. Need an option of spellcheck off while calculating WER.


     @kdavis-mozilla    "The problem is that when no training occurs global_train_time doesn't have a valid value for being passed to format_duration()" @reuben already did a pull request with resume checkpoint hot fix.  Agreed with the discussion on long term fix for decoupling WER and Training.

**Comment by kdavis-mozilla** [comment]
@bikramjitroy Do you have thoughts on the edge case in PR comment?

On your comments:

**1) There is a deadlock in util.spell "correction" API. STacktrace below**

I haven't seen this before. Could you open a new issue for it? Thanks!

**2) BatchSize more than 12 is making the training OOM.**

Yes, one has to tune the batch size to the VRAM of one's GPU. Doing so is usually a trail and error process where one aims too high and gets OOM errors.

**3) Training Time:**

The training time of about 2h's for 100h of speech using a p2.8xlarge instance is not unreasonable and is in line with what we've seen too.

**Comment by kdavis-mozilla** [comment]
@bikramjitroy As we just did a large re-factor of the code base, this PR is no longer able to merge. Also, the problem this PR was addressing should now be resolved in the refactoring. If you find it's still there, then please submit another PR.

Sorry that the re-factoring overlapped with your PR. But, non-the-less, thanks for the PR!
